crates/auto-hash-map/src/map.rs

```diff
@@ -0,0 +1,276 @@
+use std::{
+    collections::{hash_map::DefaultHasher, HashMap},
+    fmt::Debug,
+    hash::{BuildHasher, BuildHasherDefault, Hash},
+};
+
+use crate::{MAX_LIST_SIZE, MIN_HASH_SIZE};
+
+#[derive(Clone)]
+pub enum AutoMap<K, V, H = BuildHasherDefault<DefaultHasher>> {
+    List(Vec<(K, V)>),
+    Map(Box<HashMap<K, V, H>>),
+}
+
+impl<K, V, H> Default for AutoMap<K, V, H> {
+    fn default() -> Self {
+        Self::List(Default::default())
+    }
+}
+
+impl<K: Debug, V: Debug, H> Debug for AutoMap<K, V, H> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_map().entries(self.iter()).finish()
+    }
+}
+
+impl<K, V> AutoMap<K, V, BuildHasherDefault<DefaultHasher>> {
+    pub fn new() -> Self {
+        AutoMap::List(Vec::new())
+    }
+}
+
+impl<K: Eq + Hash, V, H: BuildHasher + Default> AutoMap<K, V, H> {
+    fn convert_to_map(&mut self) -> &mut HashMap<K, V, H> {
+        if let AutoMap::List(list) = self {
+            let mut map = HashMap::with_capacity_and_hasher(MAX_LIST_SIZE * 2, Default::default());
+            for (k, v) in list.drain(..) {
+                map.insert(k, v);
+            }
+            *self = AutoMap::Map(Box::new(map));
+        }
+        if let AutoMap::Map(map) = self {
+            map
+        } else {
+            unreachable!()
+        }
+    }
+
+    fn convert_to_list(&mut self) -> &mut Vec<(K, V)> {
+        if let AutoMap::Map(map) = self {
+            let mut list = Vec::with_capacity(MAX_LIST_SIZE);
+            for (k, v) in map.drain() {
+                list.push((k, v));
+            }
+            *self = AutoMap::List(list);
+        }
+        if let AutoMap::List(list) = self {
+            list
+        } else {
+            unreachable!()
+        }
+    }
+
+    pub fn insert(&mut self, key: K, value: V) -> Option<V> {
+        match self {
+            AutoMap::List(list) => {
+                for (k, v) in list.iter_mut() {
+                    if *k == key {
+                        return Some(std::mem::replace(v, value));
+                    }
+                }
+                if list.len() == MAX_LIST_SIZE {
```

Comment:
The description of `MAX_LIST_SIZE` indicates that it should convert to a map when _larger than_, not _equal to_.

```suggestion
                if list.len() > MAX_LIST_SIZE {
```

Reply:
We want to insert an item, so we need to convert it when it's at the max.